gotham/src/router/builder/single.rs

```diff
@@ -152,13 +152,24 @@ pub trait DefineSingleRoute {
     /// #   assert_eq!(response.status(), StatusCode::ACCEPTED);
     /// # }
     /// ```
-    fn to_async<H, Fut>(self, handler: H)
+    fn to_async<'s, H, Fut>(self, handler: H)
     where
         Self: Sized,
-        H: (FnOnce(State) -> Fut) + RefUnwindSafe + Copy + Send + Sync + 'static,
-        Fut: Future<Output = HandlerResult> + Send + 'static,
+        H: (FnOnce(&'_ mut State) -> Fut) + RefUnwindSafe + Copy + Send + Sync + 'static,
+        Fut: Future<Output = HandlerResult> + Send,
```

Comment:
The problem is that you can't tie the lifetime of State to that of Fut, because it has no lifetime parameters, and therefore no way to specify the required higher-ranked lifetime bounds ( https://doc.rust-lang.org/reference/trait-bounds.html#higher-ranked-trait-bounds ). If I manage to work-around this issue then it should fix https://github.com/gotham-rs/gotham/issues/431 . I am going to chase this up in the upstream rust repo here: 
https://github.com/rust-lang/rust/issues/45994

If we manage to fix this, I will also have a good solution to the ergonomics problems in https://github.com/tag1consulting/goose/pull/22